### ISSUE TITLE
Under-voltage detection (bad power supplies and USB cables)

### ISSUE THREAD

**Issue by ssvb** [longsleep/build-pine64-image#16]
Some Pine64 users have bad power supplies or bad USB cables, so they experience reboots or some other stability issues.

As a quick and dirty temporary solution, it might be possible to read the voltage level from the PMIC  (via the arisc mailbox interface), then do some simple CPU load testing in the U-Boot while gradually increasing the CPU clock frequency (utilizing more than one CPU core and NEON might be difficult though). If a significant voltage drop is detected, then boot at a low CPU clock frequency into some initrd image, which just shows something like `"Your power supply sucks, please fix this!"` message on the HDMI monitor and the UART serial console, then stops there.

Also see:
- http://irclog.whitequark.org/linux-sunxi/2016-04-25#16272824;
- https://www.raspberrypi.org/forums/viewtopic.php?t=82373

**Comment by longsleep** [comment]
Sounds like something which sound be done in mainline U-Boot. Also i am not sure if it is really worth the efford. After all the board works fine when powered properly.

**Comment by ssvb** [comment]
Yes, people having decent power supplies and cables should have no problems. It's just that the [Pine64 kickstarter campaign](https://www.kickstarter.com/projects/pine64/pine-a64-first-15-64-bit-single-board-super-comput/description) advertised that `"The PINE A64 board provides the same power of desktop computing at a mere 2.5 to 3.5 watts"`. Which means something between 500mA and 700mA at 5V. So some people could have decided that a very weak power supply of even a USB port of their PC should be good enough. That's of course none of our concern, but you can expect that some Pine64 users will be blaming your Pine64 images for poor reliability :)

As for the mainline support, the PMIC support will have to be implemented eventually.

**Comment by ThomasKaiser** [comment]
> It's just that the Pine64 kickstarter campaign advertised that "The PINE A64 board provides the same power of desktop computing at a mere 2.5 to 3.5 watts"

So basically it's the wrong power connector combined with a marketing campaign raising wrong expectations? So it might be an option that the Pine64 folks ask for paid consultantcy or coding efforts to fix their mistakes?

**Comment by texadactyl** [comment]
Good advice.  That's what I thought was happening to me.  Then, I was suspicious of my wall socket.  Eh, it turns out that my 64GB SanDisk MicroSD was intermittently failing.  Now 100% dead.  Oiy vey!
